### PR TITLE
TEBA-44: EC as Payment Method

### PR DESCRIPTION
**Pull request by NghiaDTr**

### CODE REVIEW

File: enabler/src/utils/app.utils.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+export const importEasyCreditScript = () => {
+  const currentScript = document.querySelectorAll('script#easycredit-webcomponents');
+
+  if (currentScript.length === 0) {
+    const g = document.createElement('script');
+    g.setAttribute('id', 'easycredit-webcomponents');
+    
+    const s = document.getElementsByTagName('script')[0];
```
Comment (Tung-Huynh-Shopmacher) [inline enabler/src/utils/app.utils.ts:8]: move line 8 to after line 10

File: processor/src/routes/operation.route.ts [hunk 2]
```diff
@@ -35,8 +35,8 @@ export const operationsRoute = async (fastify: FastifyInstance, opts: FastifyPlu
   fastify.get<{ Reply: WidgetEnabledResponseSchemaDTO }>(
     '/widget-enabled',
     {
-      // preHandler: [opts.jwtAuthHook.authenticate()],
-      preHandler: [],
+      preHandler: [opts.sessionHeaderAuthHook.authenticate()],
+      // preHandler: [],
```
Comment (Tung-Huynh-Shopmacher) [inline processor/src/routes/operation.route.ts:39]: remove this line

File: processor/src/services/payment.service.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  const config = readConfiguration();
+
+  const result = {
+    webShopId: config.easyCredit.webShopId,
+  };
+
+  const errors: Array<{ code: string; message: string }> = [];
+
+  if (
+    cart?.billingAddress &&
```
Comment (Tung-Huynh-Shopmacher) [inline processor/src/services/payment.service.ts:20]: what if billingAddress or shippingAddress is missing?

File: processor/src/services/payment.service.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  ) {
+    errors.push({
+      code: 'InvalidAddress',
+      message: `Liefer- und Rechnungsadresse sind identisch und in Deutschland.`,
+    });
+  }
+
+  if (cart.totalPrice.currencyCode !== 'EUR') {
+    errors.push({
+      code: 'InvalidAddress',
```
Comment (Tung-Huynh-Shopmacher) [inline processor/src/services/payment.service.ts:32]: misleading error code

File: processor/src/routes/payment.route.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+export const paymentsRoute = async (fastify: FastifyInstance, opts: FastifyPluginOptions & OperationRouteOptions) => {
+  fastify.get<{ Reply: GetOptionResponseSchemaDTO }>(
+    '/payment-method',
+    {
+      preHandler: [opts.sessionHeaderAuthHook.authenticate()],
+      schema: {
+        querystring: GetOptionQueryStringSchema,
+        response: {
+          200: GetOptionResponseSchema,
```
Comment (Tung-Huynh-Shopmacher) [inline processor/src/routes/payment.route.ts:28]: return 400 for error

**Review by Tung-Huynh-Shopmacher — CHANGES_REQUESTED** [review]

File: enabler/index.html [hunk 6]
```diff
@@ -50,14 +54,20 @@ <h1 class="h3 mb-3 font-weight-normal">Dev Site</h1>
           <script type="module">
             import { Enabler } from "/src/main.ts";
 
+            const testCardId = '5307942b-38b4-4cbc-95f5-c3ce2e386dd2';
```
Comment (Tung-Huynh-Shopmacher) [inline enabler/index.html:57]: typo

File: enabler/src/components/payment-methods/easycredit/easycredit.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  constructor(baseOptions: BaseOptions, componentOptions: ComponentOptions) {
+    super(PaymentMethod.easycredit, baseOptions, componentOptions);
+    this.baseOptions = baseOptions;
+  }
+
+
+  async mount(selector: string) {
+    importEasyCreditScript();
+
+    const response = await this.getPaymentMethod();
```
Comment (Tung-Huynh-Shopmacher) [inline enabler/src/components/payment-methods/easycredit/easycredit.ts:33]: will this every be falsy?

File: enabler/src/components/payment-methods/easycredit/easycredit.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+    if (response) {  
+      document.querySelector(selector).insertAdjacentHTML("afterbegin", this._getTemplate(response));
```
Comment (Tung-Huynh-Shopmacher) [inline enabler/src/components/payment-methods/easycredit/easycredit.ts:36]: what if this selector not existing on the DOM?

**Review by Tung-Huynh-Shopmacher — CHANGES_REQUESTED** [review]

File: processor/src/services/payment.service.ts (hunk 3 above)
Comment (NghiaDTr) [reply processor/src/services/payment.service.ts]: I added 2 more checks for those objects, please help me to have a look anh

File: processor/src/routes/payment.route.ts (hunk 5 above)
Comment (NghiaDTr) [reply processor/src/routes/payment.route.ts]: Yes anh, I added it

File: enabler/index.html (hunk 6 above)
Comment (NghiaDTr) [reply enabler/index.html]: Sorry anh, I fixed it

File: enabler/src/components/payment-methods/easycredit/easycredit.ts (hunk 8 above)
Comment (NghiaDTr) [reply enabler/src/components/payment-methods/easycredit/easycredit.ts]: I updated it, we would throw an error, then log it to the console tab if it happens, then there will be nothing rendered

**Review by Tung-Huynh-Shopmacher — APPROVED** [review]

**Pull request merged**